Add padding='same' mode to conv{1,2,3}d

peterbell10 opened the pull request:
First part of #3867 (Pooling operators still to do)

This adds a `padding='same'` mode to the interface of `conv{n}d`and `nn.Conv{n}d`. This should match the behaviour of `tensorflow`. I couldn't find it explicitly documented but through experimentation I found `tensorflow` returns the shape `ceil(len/stride)` and always adds any extra asymmetric padding onto the right side of the input.

Since the `native_functions.yaml` schema doesn't seem to support strings or enums, I've moved the function interface into python and it now dispatches between the numerically padded `conv{n}d` and the `_conv{n}d_same` variant. Underscores because I couldn't see any way to avoid exporting a function into the `torch` namespace. 

A note on asymmetric padding. The total padding required can be odd if both the kernel-length is even  and the dilation is odd. mkldnn has native support for asymmetric padding, so there is no overhead there, but for other backends I resort to padding the input tensor by 1 on the right hand side to make the remaining padding symmetrical. In these cases, I use `TORCH_WARN_ONCE` to notify the user of the performance implications.

Stack from [ghstack](https://github.com/ezyang/ghstack):
* #52631 WIP: Support strided convolutions with padding='same'
* #45668 Use asymmetric mkldnn convolutions for padding='same' convolutions
* **#45667 Add padding='same' mode to conv{1,2,3}d**

Differential Revision: [D27170744](https://our.internmc.facebook.com/intern/diff/D27170744)

peterbell10 committed:
Add padding='same' mode to conv{1,2,3}d

[ghstack-poisoned]

peterbell10 committed:
Update on "Add padding='same' mode to conv{1,2,3}d"

First part of #3867 (Pooling operators still to do)

This adds a `padding='same'` mode to the interface of `conv{n}d`and `nn.Conv{n}d`. This should match the behaviour of `tensorflow`. I couldn't find it explicitly documented but through experimentation I found `tensorflow` returns the shape `ceil(len/stride)` and always adds any extra asymmetric padding onto the right side of the input.

Since the `native_functions.yaml` schema doesn't seem to support strings or enums, I've moved the function interface into python and it now dispatches between the numerically padded `conv{n}d` and the `_conv{n}d_same` variant. Underscores because I couldn't see any way to avoid exporting a function into the `torch` namespace. 

A note on asymmetric padding. The total padding required can be odd if both the kernel-length is even  and the dilation is odd. mkldnn has native support for asymmetric padding, so there is no overhead there, but for other backends I resort to padding the input tensor by 1 on the right hand side to make the remaining padding symmetrical. In these cases, I use `TORCH_WARN_ONCE` to notify the user of the performance implications.




[ghstack-poisoned]

peterbell10 committed:
Update on "Add padding='same' mode to conv{1,2,3}d"

First part of #3867 (Pooling operators still to do)

This adds a `padding='same'` mode to the interface of `conv{n}d`and `nn.Conv{n}d`. This should match the behaviour of `tensorflow`. I couldn't find it explicitly documented but through experimentation I found `tensorflow` returns the shape `ceil(len/stride)` and always adds any extra asymmetric padding onto the right side of the input.

Since the `native_functions.yaml` schema doesn't seem to support strings or enums, I've moved the function interface into python and it now dispatches between the numerically padded `conv{n}d` and the `_conv{n}d_same` variant. Underscores because I couldn't see any way to avoid exporting a function into the `torch` namespace. 

A note on asymmetric padding. The total padding required can be odd if both the kernel-length is even  and the dilation is odd. mkldnn has native support for asymmetric padding, so there is no overhead there, but for other backends I resort to padding the input tensor by 1 on the right hand side to make the remaining padding symmetrical. In these cases, I use `TORCH_WARN_ONCE` to notify the user of the performance implications.




[ghstack-poisoned]

peterbell10 committed:
Update on "Add padding='same' mode to conv{1,2,3}d"

First part of #3867 (Pooling operators still to do)

This adds a `padding='same'` mode to the interface of `conv{n}d`and `nn.Conv{n}d`. This should match the behaviour of `tensorflow`. I couldn't find it explicitly documented but through experimentation I found `tensorflow` returns the shape `ceil(len/stride)` and always adds any extra asymmetric padding onto the right side of the input.

Since the `native_functions.yaml` schema doesn't seem to support strings or enums, I've moved the function interface into python and it now dispatches between the numerically padded `conv{n}d` and the `_conv{n}d_same` variant. Underscores because I couldn't see any way to avoid exporting a function into the `torch` namespace. 

A note on asymmetric padding. The total padding required can be odd if both the kernel-length is even  and the dilation is odd. mkldnn has native support for asymmetric padding, so there is no overhead there, but for other backends I resort to padding the input tensor by 1 on the right hand side to make the remaining padding symmetrical. In these cases, I use `TORCH_WARN_ONCE` to notify the user of the performance implications.




[ghstack-poisoned]

peterbell10 committed:
Update on "Add padding='same' mode to conv{1,2,3}d"

First part of #3867 (Pooling operators still to do)

This adds a `padding='same'` mode to the interface of `conv{n}d`and `nn.Conv{n}d`. This should match the behaviour of `tensorflow`. I couldn't find it explicitly documented but through experimentation I found `tensorflow` returns the shape `ceil(len/stride)` and always adds any extra asymmetric padding onto the right side of the input.

Since the `native_functions.yaml` schema doesn't seem to support strings or enums, I've moved the function interface into python and it now dispatches between the numerically padded `conv{n}d` and the `_conv{n}d_same` variant. Underscores because I couldn't see any way to avoid exporting a function into the `torch` namespace. 

A note on asymmetric padding. The total padding required can be odd if both the kernel-length is even  and the dilation is odd. mkldnn has native support for asymmetric padding, so there is no overhead there, but for other backends I resort to padding the input tensor by 1 on the right hand side to make the remaining padding symmetrical. In these cases, I use `TORCH_WARN_ONCE` to notify the user of the performance implications.




[ghstack-poisoned]

peterbell10 committed:
Update on "Add padding='same' mode to conv{1,2,3}d"

First part of #3867 (Pooling operators still to do)

This adds a `padding='same'` mode to the interface of `conv{n}d`and `nn.Conv{n}d`. This should match the behaviour of `tensorflow`. I couldn't find it explicitly documented but through experimentation I found `tensorflow` returns the shape `ceil(len/stride)` and always adds any extra asymmetric padding onto the right side of the input.

Since the `native_functions.yaml` schema doesn't seem to support strings or enums, I've moved the function interface into python and it now dispatches between the numerically padded `conv{n}d` and the `_conv{n}d_same` variant. Underscores because I couldn't see any way to avoid exporting a function into the `torch` namespace. 

A note on asymmetric padding. The total padding required can be odd if both the kernel-length is even  and the dilation is odd. mkldnn has native support for asymmetric padding, so there is no overhead there, but for other backends I resort to padding the input tensor by 1 on the right hand side to make the remaining padding symmetrical. In these cases, I use `TORCH_WARN_ONCE` to notify the user of the performance implications.




[ghstack-poisoned]

peterbell10 committed:
Update on "Add padding='same' mode to conv{1,2,3}d"

First part of #3867 (Pooling operators still to do)

This adds a `padding='same'` mode to the interface of `conv{n}d`and `nn.Conv{n}d`. This should match the behaviour of `tensorflow`. I couldn't find it explicitly documented but through experimentation I found `tensorflow` returns the shape `ceil(len/stride)` and always adds any extra asymmetric padding onto the right side of the input.

Since the `native_functions.yaml` schema doesn't seem to support strings or enums, I've moved the function interface into python and it now dispatches between the numerically padded `conv{n}d` and the `_conv{n}d_same` variant. Underscores because I couldn't see any way to avoid exporting a function into the `torch` namespace. 

A note on asymmetric padding. The total padding required can be odd if both the kernel-length is even  and the dilation is odd. mkldnn has native support for asymmetric padding, so there is no overhead there, but for other backends I resort to padding the input tensor by 1 on the right hand side to make the remaining padding symmetrical. In these cases, I use `TORCH_WARN_ONCE` to notify the user of the performance implications.




[ghstack-poisoned]

peterbell10 committed:
Update on "Add padding='same' mode to conv{1,2,3}d"

First part of #3867 (Pooling operators still to do)

This adds a `padding='same'` mode to the interface of `conv{n}d`and `nn.Conv{n}d`. This should match the behaviour of `tensorflow`. I couldn't find it explicitly documented but through experimentation I found `tensorflow` returns the shape `ceil(len/stride)` and always adds any extra asymmetric padding onto the right side of the input.

Since the `native_functions.yaml` schema doesn't seem to support strings or enums, I've moved the function interface into python and it now dispatches between the numerically padded `conv{n}d` and the `_conv{n}d_same` variant. Underscores because I couldn't see any way to avoid exporting a function into the `torch` namespace. 

A note on asymmetric padding. The total padding required can be odd if both the kernel-length is even  and the dilation is odd. mkldnn has native support for asymmetric padding, so there is no overhead there, but for other backends I resort to padding the input tensor by 1 on the right hand side to make the remaining padding symmetrical. In these cases, I use `TORCH_WARN_ONCE` to notify the user of the performance implications.




[ghstack-poisoned]

peterbell10 committed:
Update on "Add padding='same' mode to conv{1,2,3}d"

First part of #3867 (Pooling operators still to do)

This adds a `padding='same'` mode to the interface of `conv{n}d`and `nn.Conv{n}d`. This should match the behaviour of `tensorflow`. I couldn't find it explicitly documented but through experimentation I found `tensorflow` returns the shape `ceil(len/stride)` and always adds any extra asymmetric padding onto the right side of the input.

Since the `native_functions.yaml` schema doesn't seem to support strings or enums, I've moved the function interface into python and it now dispatches between the numerically padded `conv{n}d` and the `_conv{n}d_same` variant. Underscores because I couldn't see any way to avoid exporting a function into the `torch` namespace. 

A note on asymmetric padding. The total padding required can be odd if both the kernel-length is even  and the dilation is odd. mkldnn has native support for asymmetric padding, so there is no overhead there, but for other backends I resort to padding the input tensor by 1 on the right hand side to make the remaining padding symmetrical. In these cases, I use `TORCH_WARN_ONCE` to notify the user of the performance implications.




[ghstack-poisoned]

peterbell10 committed:
Update on "Add padding='same' mode to conv{1,2,3}d"

First part of #3867 (Pooling operators still to do)

This adds a `padding='same'` mode to the interface of `conv{n}d`and `nn.Conv{n}d`. This should match the behaviour of `tensorflow`. I couldn't find it explicitly documented but through experimentation I found `tensorflow` returns the shape `ceil(len/stride)` and always adds any extra asymmetric padding onto the right side of the input.

Since the `native_functions.yaml` schema doesn't seem to support strings or enums, I've moved the function interface into python and it now dispatches between the numerically padded `conv{n}d` and the `_conv{n}d_same` variant. Underscores because I couldn't see any way to avoid exporting a function into the `torch` namespace. 

A note on asymmetric padding. The total padding required can be odd if both the kernel-length is even  and the dilation is odd. mkldnn has native support for asymmetric padding, so there is no overhead there, but for other backends I resort to padding the input tensor by 1 on the right hand side to make the remaining padding symmetrical. In these cases, I use `TORCH_WARN_ONCE` to notify the user of the performance implications.




[ghstack-poisoned]

peterbell10 committed:
Update on "Add padding='same' mode to conv{1,2,3}d"

First part of #3867 (Pooling operators still to do)

This adds a `padding='same'` mode to the interface of `conv{n}d`and `nn.Conv{n}d`. This should match the behaviour of `tensorflow`. I couldn't find it explicitly documented but through experimentation I found `tensorflow` returns the shape `ceil(len/stride)` and always adds any extra asymmetric padding onto the right side of the input.

Since the `native_functions.yaml` schema doesn't seem to support strings or enums, I've moved the function interface into python and it now dispatches between the numerically padded `conv{n}d` and the `_conv{n}d_same` variant. Underscores because I couldn't see any way to avoid exporting a function into the `torch` namespace. 

A note on asymmetric padding. The total padding required can be odd if both the kernel-length is even  and the dilation is odd. mkldnn has native support for asymmetric padding, so there is no overhead there, but for other backends I resort to padding the input tensor by 1 on the right hand side to make the remaining padding symmetrical. In these cases, I use `TORCH_WARN_ONCE` to notify the user of the performance implications.




[ghstack-poisoned]

peterbell10 commented:
@gchanan are you available to review this?

mruberry commented:
FYI @jbschlosser, @ngimel and I are planning on looking at this together early next week.

mruberry commented on torch/nn/functional.py:
```diff
@@ -63,13 +72,23 @@
     {cudnn_reproducibility_note}
 """.format(**reproducibility_notes, **tf32_notes) + r"""
+
+    Padding mode 'valid' is the same as no padding. 'same' pads the input so
```
Let's be careful to refer to "padding" vs. "padding mode", since even though "padding_mode" is not a parameter of the functional convolution it is a parameter of the module, and we don't want users to confuse the two.

mruberry replied:
The start of this paragraph really jumps into to explain some padding options. Either start by explaining the padding parameter a little more, or include the "valid" vs. "same" padding effect in the args discussion below.

mruberry commented on torch/nn/functional.py:
```diff
@@ -63,13 +72,23 @@
     stride: the stride of the convolving kernel. Can be a single number or a
       tuple `(sH, sW)`. Default: 1
-    padding: implicit paddings on both sides of the input. Can be a
+    padding: implicit paddings on both sides of the input. Can be a mode {'valid', 'same'},
```
Is there a good paper reference for valid and/or same padding?

jbschlosser commented on test/test_nn.py:
```diff
@@ -888,6 +888,81 @@ def test_invalid_conv3d(self):
+
+        # non-zero padding modes require explicit padding
+        module = nn.Conv3d(in_channels=1, out_channels=1, kernel_size=(2, 3, 4),
+                           padding='same', padding_mode='circular')
```
nit: I see tests for `circular`, `reflect`, and (implicitly) `zeros` padding modes, but none for `replicate`. Please add at least one test that covers the `replicate` padding mode with `same` padding.

jbschlosser commented on torch/nn/functional.py:
```diff
@@ -30,13 +30,22 @@
+    .. warning::
+        For ``padding='same'``, if the ``weight`` is even-length and
+        ``dilation`` is odd in any dimension, a full :func:`pad` operation
+        may be needed internally. Lowering performance.
```
nit: "may be needed internally. Lowering performance." -> "will be needed internally, lowering performance."

Is "even-length weight" & "odd dilation" the only case that requires a full pad operation? If so, please update the `TORCH_WARN_ONCE` text to specify this.

peterbell10 replied:
mkldnn supports asymmetric padding, so whether or not it happens is an implementation detail. See next PR in the stack which takes advantage of that.

jbschlosser commented on torch/nn/functional.py:
```diff
@@ -63,13 +72,23 @@
+    .. warning::
+        For ``padding='same'``, if the ``weight`` is even-length and
+        ``dilation`` is odd in any dimension, a full :func:`pad` operation
+        may be needed internally. Lowering performance.
```
Same comments here as for the other docstrings.

> nit: "may be needed internally. Lowering performance." -> "will be needed internally, lowering performance."

> Is "even-length weight" & "odd dilation" the only case that requires a full pad operation? If so, please update the TORCH_WARN_ONCE text to specify this.

jbschlosser commented on torch/nn/functional.py:
```diff
@@ -98,13 +117,22 @@
+    .. warning::
+        For ``padding='same'``, if the ``weight`` is even-length and
+        ``dilation`` is odd in any dimension, a full :func:`pad` operation
+        may be needed internally. Lowering performance.
```
Same comments here as for the other docstrings.

> nit: "may be needed internally. Lowering performance." -> "will be needed internally, lowering performance."

> Is "even-length weight" & "odd dilation" the only case that requires a full pad operation? If so, please update the TORCH_WARN_ONCE text to specify this.

jbschlosser commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -1245,6 +1248,18 @@
 - func: conv3d(Tensor input, Tensor weight, Tensor? bias=None, int[3] stride=1, int[3] padding=0, int[3] dilation=1, int groups=1) -> Tensor
   use_c10_dispatcher: hacky_wrapper_for_legacy_signatures
 
+- func: conv1d.padding_mode(Tensor input, Tensor weight, Tensor? bias=None, int[1] stride=1, str padding="valid", int[1] dilation=1, int groups=1) -> Tensor
```
As mentioned in some other comments, the "padding_mode" terminology here is a bit confusing since "padding mode" is used in Conv1d / Conv2d / Conv3d modules to describe the padding fill values (e.g. zeros, reflect, etc.).

Maybe this can be changed to just `padding`?

jbschlosser commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -1245,6 +1248,18 @@
+  use_c10_dispatcher: hacky_wrapper_for_legacy_signatures
+  cpp_no_default_args: ['bias', 'stride', 'padding']
+
+- func: conv2d.padding_mode(Tensor input, Tensor weight, Tensor? bias=None, int[2] stride=1, str padding="valid", int[2] dilation=1, int groups=1) -> Tensor
```
Similar comment here as above: padding_mode -> padding.

jbschlosser commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -1245,6 +1248,18 @@
+  use_c10_dispatcher: hacky_wrapper_for_legacy_signatures
+  cpp_no_default_args: ['bias', 'stride', 'padding']
+
+- func: conv3d.padding_mode(Tensor input, Tensor weight, Tensor? bias=None, int[3] stride=1, str padding="valid", int[3] dilation=1, int groups=1) -> Tensor
```
Similar comment here as above: padding_mode -> padding.

jbschlosser reviewed (changes requested):
Looks great overall! Few minor comments re: docs / testing.

Also, please add a few entries to `new_module_tests` in `torch/testing/_internal/common_nn.py` for testing `Conv1d` / `Conv2d` / `Conv3d` with `same` padding. (That’ll do Jacobian checks)

ppwwyyxx commented:
> This should match the behaviour of tensorflow. I couldn't find it explicitly documented

The current version does not match the behavior of tensorflow, e.g. try input=224, kernel=7, stride=2 using [this tensorflow code](https://github.com/tensorflow/tensorflow/blob/3c3c0481ec087aca4fa875d6d936f19b31191fc1/tensorflow/core/framework/common_shape_fns.cc#L40-L48). There used to be a doc in tensorflow describing it, but now missing (https://github.com/tensorflow/tensorflow/issues/46980).

However, as multiple comments [1](https://github.com/pytorch/pytorch/issues/3867#issuecomment-348871623), [2](https://github.com/pytorch/pytorch/issues/3867#issuecomment-604277892), [3](https://github.com/pytorch/pytorch/issues/3867#issuecomment-774615556) pointed out, the size-dependent "same" padding calculation in tensorflow is a bad default. As I've shown in https://github.com/tensorflow/tensorflow/issues/18213, many of Google research's own code implements their own "better" same padding than tensorflow. Matching tensorflow does not necessarily have to be our goal.

peterbell10 commented:
>The current version does not match the behavior of tensorflow, e.g. try input=224, kernel=7, stride=2 using this tensorflow code. There used to be a doc in tensorflow describing it, but now missing (tensorflow/tensorflow#46980).

Interesting. The TF calculation is a bit verbose, but it seems to be calculating:
```
total_padding = max(0, dilation * (kernel_size - 1) - (input_size - 1) % stride)
```
compared to this PR which calculates:
```
total_padding = dilation * (kernel_size - 1)
```

This PR uses a consistent padding amount, regardless of input size. Whereas TF is using the minimum padding required for a given input size. Both result in the same output shape, and in the absence of strides also the same result. However, if the convolution has strides >2 then TF's convention means the edge values came from on average more image elements and fewer padding elements.

I can see arguments for both ways; perhaps it would make sense to have both options?

jbschlosser commented:
> This PR uses a consistent padding amount, regardless of input size. Whereas TF is using the minimum padding required for a given input size.

@peterbell10 I think this is not quite true- the "wiggle room" logic makes the padding amount input size dependent for this PR as well. To avoid the performance / correctness issues of input size dependent padding brought up by @ppwwyyxx, this would have to change.

peterbell10 commented:
>>This PR uses a consistent padding amount, regardless of input size. Whereas TF is using the minimum padding required for a given input size.

>@peterbell10 I think this is not quite true- the "wiggle room" logic makes the padding amount input size dependent for this PR as well. To avoid the performance / correctness issues of input size dependent padding brought up by @ppwwyyxx, this would have to change.

The "wiggle room" only every removes padding from the right side of the tensor. If you consider that the output size is the same, then you can reason that the output is unaffected. So, from the user's perspective this should not matter. It's just a performance optimization.

bilalsal commented:
IMO, if we restrict our support of SAME padding to stride=1, we provide 90% of convenience our users want, and mitigate the problem of input-dependence.
For Conv layers that perform downsampling (stride > 1), the output anyhow cannot be SAME size as the input. We can just fail with a message that helps the user makes an informed choice:
- Is padding needed here (e.g. in case of semseg, to avoid drifts in input-output pixel matching)?
- Would they be OK with 'replicate', 'reflect', 'circular' as safer alternatives to the default zero in this case (they do not need to disproportional zeros)?
- Would AvgPool / MaxPool work instead of strided convolution (they are rigid operations, and won't be skewed by disproportional zeros)?

jbschlosser commented:
> IMO, if we restrict our support of SAME padding to stride=1, we provide 90% of the convenience our users want, and mitigate the problem of input-dependence.

As described by @peterbell10 above, the current approach in this PR uses a consistent padding amount, regardless of input size, and thus does not suffer from the problems of input-dependence. So it's unclear to me what the `stride = 1` restriction would accomplish.

Note that asymmetric padding is possible even when `stride = 1`, specifically when the kernel length is even and dilation is odd. To me, it makes more sense to restrict or warn for that case, rather than for `stride > 1`.

Is there something uniquely dangerous about the asymmetric padding + `stride > 1` case that I missed, even if padding is not input-dependent? @bilalsal

ngimel commented:
If we use input size independent padding then
1) as @peterbell10 pointed out, we would, on average, use more padded pixels and fewer image pixels than tensorflow. 
2) results would be different from tensorflow even for configs where the number of padded pixels used in computation is the same
Given that the stated goal is being compatible with tensorflow, I think those inconsistencies are problematic.

jbschlosser commented:
> If we use input size independent padding then
> 1) as @peterbell10 pointed out, we would, on average, use more padded pixels and fewer image pixels than tensorflow. 
> 2) results would be different from tensorflow even for configs where the number of padded pixels used in computation is the same
> Given that the stated goal is being compatible with tensorflow, I think those inconsistencies are problematic. 

Fair enough. I wonder what % of use cases would be satisfied with the "conceptual" same padding in the PR. Certainly this would be a superset of use cases involving `stride = 1` (as semantics are exactly equal to TF's for this case). For those that still desire the _exact_ semantics of TF, manual padding is an option.

Of course, if the majority of users requesting 'same' padding need those _exact_ semantics, "conceptual" same is not good enough, but I'm skeptical that they do and I wonder if we could get more information on this.

peterbell10 committed:
Update on "Add padding='same' mode to conv{1,2,3}d"

First part of #3867 (Pooling operators still to do)

This adds a `padding='same'` mode to the interface of `conv{n}d`and `nn.Conv{n}d`. This should match the behaviour of `tensorflow`. I couldn't find it explicitly documented but through experimentation I found `tensorflow` returns the shape `ceil(len/stride)` and always adds any extra asymmetric padding onto the right side of the input.

Since the `native_functions.yaml` schema doesn't seem to support strings or enums, I've moved the function interface into python and it now dispatches between the numerically padded `conv{n}d` and the `_conv{n}d_same` variant. Underscores because I couldn't see any way to avoid exporting a function into the `torch` namespace. 

A note on asymmetric padding. The total padding required can be odd if both the kernel-length is even  and the dilation is odd. mkldnn has native support for asymmetric padding, so there is no overhead there, but for other backends I resort to padding the input tensor by 1 on the right hand side to make the remaining padding symmetrical. In these cases, I use `TORCH_WARN_ONCE` to notify the user of the performance implications.




[ghstack-poisoned]

peterbell10 commented on aten/src/ATen/native/Convolution.cpp:
```diff
@@ -561,6 +562,107 @@ at::Tensor conv3d(
+    return at::native::convolution(
+        input, weight, bias, stride, padding_, dilation, false, padding_, groups);
+  }
+  TORCH_CHECK(false, "Invalid padding mode '", padding, "'");
```
Changed "padding mode" to "padding string" throughout.

peterbell10 committed:
Update on "Add padding='same' mode to conv{1,2,3}d"

First part of #3867 (Pooling operators still to do)

This adds a `padding='same'` mode to the interface of `conv{n}d`and `nn.Conv{n}d`. This should match the behaviour of `tensorflow`. I couldn't find it explicitly documented but through experimentation I found `tensorflow` returns the shape `ceil(len/stride)` and always adds any extra asymmetric padding onto the right side of the input.

Since the `native_functions.yaml` schema doesn't seem to support strings or enums, I've moved the function interface into python and it now dispatches between the numerically padded `conv{n}d` and the `_conv{n}d_same` variant. Underscores because I couldn't see any way to avoid exporting a function into the `torch` namespace. 

A note on asymmetric padding. The total padding required can be odd if both the kernel-length is even  and the dilation is odd. mkldnn has native support for asymmetric padding, so there is no overhead there, but for other backends I resort to padding the input tensor by 1 on the right hand side to make the remaining padding symmetrical. In these cases, I use `TORCH_WARN_ONCE` to notify the user of the performance implications.




[ghstack-poisoned]

peterbell10 commented:
@jbschlosser, @mruberry PTAL. I've addressed most of your comments and also fixed the C++ `torch::nn` API.

Also, regarding strided convolutions:
>IMO, if we restrict our support of SAME padding to stride=1

Should we consider this as a first step so this PR can move ahead without exact TF compatibility?

jbschlosser commented:
> > IMO, if we restrict our support of SAME padding to stride=1
> 
> Should we consider this as a first step so this PR can move ahead without exact TF compatibility?

Yes, I think that's the way to go, since stride=1 support is uncontroversial. I suggest we get that in and open a new issue for same padding w/ stride > 1 support.

For stride > 1, I think we should support multiple sets of semantics (e.g. SAME, SAME_TF, etc.; names TBD) for max flexibility and try to discourage users from the TF case with doc warnings. If we do this, the naming needs to be nailed down, and pushing it to a later issue prevents this PR from being held up.

Beyond the fixed padding used in this PR and TF's semantics, I think the padding scheme proposed in Yangqing's [notebook](https://gist.github.com/Yangqing/47772de7eb3d5dbbff50ffb0d7a98964) is a third option: "If we pad evenly by ceil(pad_total / 2) where / is float division, we get exactly the same output size, and a much better padding mechanism."

peterbell10 committed:
Update on "Add padding='same' mode to conv{1,2,3}d"

First part of #3867 (Pooling operators still to do)

This adds a `padding='same'` mode to the interface of `conv{n}d`and `nn.Conv{n}d`. This should match the behaviour of `tensorflow`. I couldn't find it explicitly documented but through experimentation I found `tensorflow` returns the shape `ceil(len/stride)` and always adds any extra asymmetric padding onto the right side of the input.

Since the `native_functions.yaml` schema doesn't seem to support strings or enums, I've moved the function interface into python and it now dispatches between the numerically padded `conv{n}d` and the `_conv{n}d_same` variant. Underscores because I couldn't see any way to avoid exporting a function into the `torch` namespace. 

A note on asymmetric padding. The total padding required can be odd if both the kernel-length is even  and the dilation is odd. mkldnn has native support for asymmetric padding, so there is no overhead there, but for other backends I resort to padding the input tensor by 1 on the right hand side to make the remaining padding symmetrical. In these cases, I use `TORCH_WARN_ONCE` to notify the user of the performance implications.




[ghstack-poisoned]

jbschlosser reviewed (commented):
Sorry for the delay- finally took the time to dig through this today in extreme detail.

Logic is correct, support is correctly restricted to `stride=1` for this PR, test coverage for the python side looks solid.

One request: since the padding mode handling logic is duplicated on the C++ module side, I think it'd be valuable to validate the various cases (valid padding, same padding + each padding mode, error case of same padding with stride > 1) with C++ module tests in `test/cpp/api/modules.cpp`.

peterbell10 committed:
Update on "Add padding='same' mode to conv{1,2,3}d"

First part of #3867 (Pooling operators still to do)

This adds a `padding='same'` mode to the interface of `conv{n}d`and `nn.Conv{n}d`. This should match the behaviour of `tensorflow`. I couldn't find it explicitly documented but through experimentation I found `tensorflow` returns the shape `ceil(len/stride)` and always adds any extra asymmetric padding onto the right side of the input.

Since the `native_functions.yaml` schema doesn't seem to support strings or enums, I've moved the function interface into python and it now dispatches between the numerically padded `conv{n}d` and the `_conv{n}d_same` variant. Underscores because I couldn't see any way to avoid exporting a function into the `torch` namespace. 

A note on asymmetric padding. The total padding required can be odd if both the kernel-length is even  and the dilation is odd. mkldnn has native support for asymmetric padding, so there is no overhead there, but for other backends I resort to padding the input tensor by 1 on the right hand side to make the remaining padding symmetrical. In these cases, I use `TORCH_WARN_ONCE` to notify the user of the performance implications.




[ghstack-poisoned]

peterbell10 committed:
Update on "Add padding='same' mode to conv{1,2,3}d"

First part of #3867 (Pooling operators still to do)

This adds a `padding='same'` mode to the interface of `conv{n}d`and `nn.Conv{n}d`. This should match the behaviour of `tensorflow`. I couldn't find it explicitly documented but through experimentation I found `tensorflow` returns the shape `ceil(len/stride)` and always adds any extra asymmetric padding onto the right side of the input.

Since the `native_functions.yaml` schema doesn't seem to support strings or enums, I've moved the function interface into python and it now dispatches between the numerically padded `conv{n}d` and the `_conv{n}d_same` variant. Underscores because I couldn't see any way to avoid exporting a function into the `torch` namespace. 

A note on asymmetric padding. The total padding required can be odd if both the kernel-length is even  and the dilation is odd. mkldnn has native support for asymmetric padding, so there is no overhead there, but for other backends I resort to padding the input tensor by 1 on the right hand side to make the remaining padding symmetrical. In these cases, I use `TORCH_WARN_ONCE` to notify the user of the performance implications.




[ghstack-poisoned]

peterbell10 committed:
Update on "Add padding='same' mode to conv{1,2,3}d"

First part of #3867 (Pooling operators still to do)

This adds a `padding='same'` mode to the interface of `conv{n}d`and `nn.Conv{n}d`. This should match the behaviour of `tensorflow`. I couldn't find it explicitly documented but through experimentation I found `tensorflow` returns the shape `ceil(len/stride)` and always adds any extra asymmetric padding onto the right side of the input.

Since the `native_functions.yaml` schema doesn't seem to support strings or enums, I've moved the function interface into python and it now dispatches between the numerically padded `conv{n}d` and the `_conv{n}d_same` variant. Underscores because I couldn't see any way to avoid exporting a function into the `torch` namespace. 

A note on asymmetric padding. The total padding required can be odd if both the kernel-length is even  and the dilation is odd. mkldnn has native support for asymmetric padding, so there is no overhead there, but for other backends I resort to padding the input tensor by 1 on the right hand side to make the remaining padding symmetrical. In these cases, I use `TORCH_WARN_ONCE` to notify the user of the performance implications.




[ghstack-poisoned]

peterbell10 committed:
Update on "Add padding='same' mode to conv{1,2,3}d"

First part of #3867 (Pooling operators still to do)

This adds a `padding='same'` mode to the interface of `conv{n}d`and `nn.Conv{n}d`. This should match the behaviour of `tensorflow`. I couldn't find it explicitly documented but through experimentation I found `tensorflow` returns the shape `ceil(len/stride)` and always adds any extra asymmetric padding onto the right side of the input.

Since the `native_functions.yaml` schema doesn't seem to support strings or enums, I've moved the function interface into python and it now dispatches between the numerically padded `conv{n}d` and the `_conv{n}d_same` variant. Underscores because I couldn't see any way to avoid exporting a function into the `torch` namespace. 

A note on asymmetric padding. The total padding required can be odd if both the kernel-length is even  and the dilation is odd. mkldnn has native support for asymmetric padding, so there is no overhead there, but for other backends I resort to padding the input tensor by 1 on the right hand side to make the remaining padding symmetrical. In these cases, I use `TORCH_WARN_ONCE` to notify the user of the performance implications.




[ghstack-poisoned]

jbschlosser reviewed (approved):
lgtm! could you please rebase to fix the merge conflicts?

peterbell10 committed:
Update on "Add padding='same' mode to conv{1,2,3}d"

First part of #3867 (Pooling operators still to do)

This adds a `padding='same'` mode to the interface of `conv{n}d`and `nn.Conv{n}d`. This should match the behaviour of `tensorflow`. I couldn't find it explicitly documented but through experimentation I found `tensorflow` returns the shape `ceil(len/stride)` and always adds any extra asymmetric padding onto the right side of the input.

Since the `native_functions.yaml` schema doesn't seem to support strings or enums, I've moved the function interface into python and it now dispatches between the numerically padded `conv{n}d` and the `_conv{n}d_same` variant. Underscores because I couldn't see any way to avoid exporting a function into the `torch` namespace. 

A note on asymmetric padding. The total padding required can be odd if both the kernel-length is even  and the dilation is odd. mkldnn has native support for asymmetric padding, so there is no overhead there, but for other backends I resort to padding the input tensor by 1 on the right hand side to make the remaining padding symmetrical. In these cases, I use `TORCH_WARN_ONCE` to notify the user of the performance implications.




[ghstack-poisoned]

peterbell10 commented:
Fixed merge conflicts, CI failure is unrelated:
```
Mar 17 16:05:40 + pip install -q --user git+https://github.com/pytorch/vision.git@8a2dc6f22ac4389ccba8859aa1e1cb14f1ee53db
Mar 17 16:07:53   ERROR: Command errored out with exit status 128:
Mar 17 16:07:53    command: git fetch -q https://github.com/pytorch/vision.git 8a2dc6f22ac4389ccba8859aa1e1cb14f1ee53db
Mar 17 16:07:53        cwd: /tmp/pip-req-build-jueuvyvq
Mar 17 16:07:53   Complete output (1 lines):
Mar 17 16:07:53   fatal: unable to access 'https://github.com/pytorch/vision.git/': Failed to connect to github.com port 443: Connection timed out
```

xwang233 commented on torch/nn/modules/conv.py:
```diff
@@ -986,6 +1029,7 @@ def forward(self, input: Tensor, output_size: Optional[List[int]] = None) -> Ten
         if self.padding_mode != 'zeros':
             raise ValueError('Only `zeros` padding mode is supported for ConvTranspose3d')
 
+        assert isinstance(self.padding, tuple)
```
I think this assertion is wrong. The following code will crash because padding is `list`, not `tuple`

```python
import torch

x = torch.randn(4, 8, 32, 32, device='cuda', dtype=torch.float)
net = torch.nn.ConvTranspose2d(8, 16, kernel_size=3, padding=[3, 3])
net = net.cuda()

y = net(x)
torch.cuda.synchronize()
```

error message:

```
Traceback (most recent call last):
  File "/home/xwang/Developer/pytorch-test/random-things/padding-tuple.py", line 7, in <module>
    y = net(x)
  File "/home/xwang/Developer/pytorch/torch/nn/modules/module.py", line 1015, in _call_impl
    return forward_call(*input, **kwargs)
  File "/home/xwang/Developer/pytorch/torch/nn/modules/conv.py", line 890, in forward
    assert isinstance(self.padding, tuple)
AssertionError
```

Can you please fix it? @peterbell10

ngimel replied:
@xwang233 thanks for pointing out! Can you please file a separate issue?
